mlflow/java/client/src/main/java/org/mlflow/artifacts/ArtifactRepository.java

```diff
@@ -0,0 +1,103 @@
+  /**
+   * Uploads the given local file to the run's root artifact directory. For example,
+   *
+   *  logArtifact("/my/localModel")
+   *  listArtifacts() // returns "localModel"
+   *
+   * @param localFile File to upload. Must exist, and must be a simple file (not a directory).
+   */
+  void logArtifact(File localFile);
+
+  /**
+   * Uploads the given local file to an artifactPath within the run's root directory. For example,
+   *
+   *   logArtifact("/my/localModel", "model")
+   *   listArtifacts("model") // returns "model/localModel"
+   *
+   * (i.e., the localModel file is now available in model/localModel).
+   *
+   * @param localFile File to upload. Must exist, and must be a simple file (not a directory).
+   * @param artifactPath Artifact path relative to the run's root directory. Should NOT
+   *                     start with a /.
+   */
+  void logArtifact(File localFile, @Nonnull String artifactPath);
+
+  /**
+   * Uploads all files within the given local director the run's root artifact directory.
+   * For example, if /my/local/dir/ contains two files "file1" and "file2", then
+   *
+   *  logArtifacts("/my/local/dir")
+   *  listArtifacts() // returns "file1" and "file2"
+   *
+   * @param localDir Directory to upload. Must exist, and must be a directory (not a simple file).
+   */
+  void logArtifacts(File localDir);
+
+
+  /**
+   * Uploads all files within the given local director an artifactPath within the run's root
+   * artifact directory. For example, if /my/local/dir/ contains two files "file1" and "file2", then
+   *
+   *  logArtifacts("/my/local/dir", "model")
+   *  listArtifacts("model") // returns "model/file1" and "model/file2"
+   *
+   * (i.e., the contents of the local directory are now available in model/).
+   *
+   * @param localDir Directory to upload. Must exist, and must be a directory (not a simple file).
+   * @param artifactPath Artifact path relative to the run's root directory. Should NOT
+   *                     start with a /.
+   */
+  void logArtifacts(File localDir, @Nonnull String artifactPath);
+
+  /**
+   * Lists the artifacts immediately under the run's root artifact directory. This does not
+   * recursively list; instead, it will return FileInfos with isDir=true where further
+   * listing may be done.
+   */
+  List<FileInfo> listArtifacts();
+
+  /**
+   * Lists the artifacts immediately under the given artifactPath within the run's root artifact
+   * irectory. This does not recursively list; instead, it will return FileInfos with isDir=true
+   * where further listing may be done.
+   * @param artifactPath Artifact path relative to the run's root directory. Should NOT
+   *                     start with a /.
+   */
+  List<FileInfo> listArtifacts(@Nonnull String artifactPath);
+
+  /**
+   * Returns a local directory containing *all* artifacts within the run's artifact directory.
+   * This method is recursive, and so may be an expensive operation if the artifact directory
```

Comment:
Given that this is an interface, I recommend not putting this comment. An implementation of Artifact Store might choose to hash deep paths for instance.

Reply:
What do you mean hash deep paths? I meant that we recursively download the entire directory.

Reply:
I see what you mean. The comment is a bit confusing -- suggesting that the method has to be recursive. Could we instead say

```
This will download the entire artifact directory path. This could be an expensive operation if the the artifiact directory is large and deep.
```